### DISCUSSION TITLE
FLAT Sprite anims not working when frames start after 0

### DISCUSSION THREAD

**Discussion by ericdrosas87** [enable3d/enable3d#212] [Help / Questions] [answered]
Hello, I am new to Enable3D, but not to video game dev in general. First off, great job with this framework! I understand that the 2D functionality is either experimental/in-progress/etc. but I can't seem to get my anims to work.

I'm using the [Sprite Sheet example ](https://enable3d.io/examples/flat/sprite-sheet.html)as a starting place and my intention is to create a billboard proof-of-concept with 2D sprites in a 3D world. The `idle` animation works fine, as it starts from 0 and ends on 3. The `run` and `jump` animations do not, however. In fact, it seems the `.play()` method does not seem to like whenever a frame does not start at 0.

Is this expected behavior for the current state of Enable3D? Is there a workaround other than having separate sprite sheets for different animations of the same character?

Update 1:

More information:

The root of the issue is that when the starting frame # is higher than the length of the timeline array then the _currentIndex variable is never being assigned to 0 and iterating across the timeline array. I'll update when I know more.

**Comment by ericdrosas87** [comment, marked as answer]
Found the issue, I'm going off of the minified JS file built from source map as I see it in the browser, here is the defective `_play()` function in `actionSprite`:

```javascript

    _play(name) {
        this._stop();
        this._currentAnimationName = name;
        const animation = this.getAnimationByName(name);
        if (!animation)
            console.warn(`Animation "${name}" does not exist!`);
        const { timeline, rate, repeat } = animation;
        this._currentIndex = -1;
        let loops = 0;
        const playNextFrame = () => {
            this._currentIndex++;
            if (this._currentIndex >= timeline.length) {
                this._currentIndex = 0;
                loops++;
            }
            this._currentIndex = timeline[this._currentIndex];
            const shouldStop = !(repeat === -1 || loops < repeat);
            if (shouldStop) {
                this._stop();
                this._events.emit('complete');
                return;
            }
            this.setFrame(this._currentIndex);
        };
        playNextFrame();
        this.interval = window.setInterval(() => {
            playNextFrame();
        }, 1000 / rate);
        return { onComplete: (cb) => this._events.once('complete', cb) };
    }
```

The line in question is `this._currentIndex = timeline[this._currentIndex];`, so when `playNextFrame()` is called before `this.interval = ...` then `this._currentIndex` is `-1`. Once in the interval, the first IF check passes and then `this._currentIndex` is `0`. But, immediately after `this_currentIndex` is assigned to the first value of the `timeline` array. 

The code should look like this:

```javascript
_play(t) {
    this._stop(), (this._currentAnimationName = t);
    const e = this.getAnimationByName(t);
    e || console.warn(`Animation "${t}" does not exist!`);
    const { timeline: n, rate: i, repeat: r } = e;
    this._currentIndex = -1;
    let s = 0;
    const a = () => {
        this._currentIndex++;
        if (this._currentIndex >= n.length) {
            this._currentIndex = 0;
            s++;
        } 

        const shouldStop = !(r === -1 || s < r);
        if (shouldStop) {
            this._stop();
            this._events.emit('complete');
            return;
        }
        this.setFrame(n[this._currentIndex]);
    };
    a();
    this.interval = window.setInterval(() => {
        a();
    }, 1000 / i);
    return { onComplete: (t) => this._events.once('complete', t) };
}
```
(Note the minified variables I had to use outside of the built sourcemap)

The reason why animation is working when starting at frame `0` is because the index is always equal to the corresponding value of the timeline array.

Can someone fix this? Or point me in the direction of the non-minified JS file so I can open up a PR?

**Reply by yandeu** [reply]
Hi, I'm currently on vacation. I was never able to finish the 2D elements (I call them "flat").
The source code is here: https://github.com/enable3d/enable3d/tree/master/packages/threeGraphics/src/flat
Feel free to open an Issue or PR :)

**Reply by ericdrosas87** [reply]
No worries!

>  I was never able to finish the 2D elements

Is that to say that you are no longer actively maintaining this project?

> Feel free to open an Issue or PR :)

I'll open up a PR today!

**Reply by yandeu** [reply]
> Is that to say that you are no longer actively maintaining this project?

Wrong, I am maintaining the project, but the 2d elements were never my main focus. I am happy to improve the 2d elements if someone (like you) is interested in.

> I'll open up a PR today!

Thanks 😊

---

Edit:  
_I will probably invest some more time in VR and AR for enable3d._

**Reply by ericdrosas87** [reply]
Ah understood. I'm using Enable3D to see if I can create a billboarding proof-of-concept in the style of Octopath Traveler. I'd greatly appreciate any improvements you have the time to implement. I'll probably raise issues or discussions when I come across things like this though if you don't mind?

**Reply by yandeu** [reply]
Yes, of course you can do that. I will look at all issues and try to merge PRs.

You shoud be able to realize your game with enable3d. Remember, it's just a very small wrapper around three.js. In enable3d you can use anything three.js provides.
